### ISSUE TITLE
Metric Learning - Triplet Loss

### ISSUE THREAD

**Issue by msafi04** [keras-team/keras-io#160]
I am trying to make Triplet loss from your Metric Leaning code. Below is my code

class EmbeddingModel(tf.keras.Model):
    def train_step(self, data):        
        if isinstance(data, tuple):
            data = data[0]
        anchors, positives, negatives = data[0], data[1], data[2]

        with tf.GradientTape() as tape:
            anchor_embeddings = self(anchors, training = True)
            positive_embeddings = self(positives, training = True)
            negative_embeddings = self(negatives, training = True)

            similarities_p = tf.einsum(
                "ae,pe->ap", anchor_embeddings, positive_embeddings
            )
            
            similarities_n = tf.einsum(
                "ae,ne->an", anchor_embeddings, negative_embeddings
            )

            
            temperature = 0.2
            d = K.maximum(similarities_p - similarities_n + temperature, 0)
        
            sparse_labels = tf.range(num_classes)
            loss = self.compiled_loss(sparse_labels, d)

        # Calculate gradients and apply via optimizer.
        gradients = tape.gradient(loss, self.trainable_variables)
        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))

        # Update and return metrics (specifically the one for the loss value).
        self.compiled_metrics.update_state(sparse_labels, similarities)
        return {m.name: m.result() for m in self.metrics}

After I do model.fit() (my embedding model is ResNet50, the loss doesn't seem to reduce at all

Epoch 1/10
100/100 [==============================] - 24s 244ms/step - loss: 113.5303
Epoch 2/10
100/100 [==============================] - 24s 239ms/step - loss: 106.8113
Epoch 3/10
100/100 [==============================] - 24s 245ms/step - loss: 106.8011
Epoch 4/10
100/100 [==============================] - 24s 237ms/step - loss: 103.7386
Epoch 5/10
100/100 [==============================] - 23s 232ms/step - loss: 103.6625
Epoch 6/10
100/100 [==============================] - 25s 246ms/step - loss: 103.6164
Epoch 7/10
100/100 [==============================] - 24s 238ms/step - loss: 103.6200

Please advise and help me how do I fix the issue.

**Comment by msafi04** [comment]
EDIT:

class EmbeddingModel(tf.keras.Model):
    def train_step(self, data):
        if isinstance(data, tuple):
            data = data[0]
        anchors, positives, negatives = data[0], data[1], data[2]

        with tf.GradientTape() as tape:
          
            anchor_embeddings = self(anchors, training = True)
            positive_embeddings = self(positives, training = True)
            negative_embeddings = self(negatives, training = True)

            temperature = 0.2
            
            loss_layer = TripletLossLayer(alpha = temperature, name = 'triplet_loss_layer')([anchor_embeddings, positive_embeddings, negative_embeddings])
 
           sparse_labels = tf.range(num_classes)
           loss = self.compiled_loss(sparse_labels, loss_layer)

        gradients = tape.gradient(loss, self.trainable_variables)
        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))

        self.compiled_metrics.update_state(sparse_labels, loss_layer)
        return {m.name: m.result() for m in self.metrics}

class TripletLossLayer(tf.keras.Model):
    def __init__(self, alpha, **kwargs):
        self.alpha = alpha
        super(TripletLossLayer, self).__init__(**kwargs)
    
    def triplet_loss(self, inputs):
        anchor, positive, negative = inputs
        
        p_dist = tf.einsum(
                "ae,pe->ap", anchor, positive
            )
        n_dist = tf.einsum(
                "ae,ne->an", anchor, negative
            )
        return K.sum(K.maximum(p_dist - n_dist + self.alpha, 0), axis = 0)
    
    def call(self, inputs):
        loss = self.triplet_loss(inputs)
        self.add_loss(loss)
        return loss


I repurposed @matpalm code like above, but the loss is all over the place.

Epoch 1/100
100/100 [==============================] - 27s 273ms/step - loss: 389.7484
Epoch 2/100
100/100 [==============================] - 25s 249ms/step - loss: 107.5993
Epoch 3/100
100/100 [==============================] - 25s 248ms/step - loss: 103.6164
Epoch 4/100
100/100 [==============================] - 24s 236ms/step - loss: 103.6162
Epoch 5/100
100/100 [==============================] - 23s 233ms/step - loss: 103.6163
Epoch 6/100
100/100 [==============================] - 24s 241ms/step - loss: 103.5967
Epoch 7/100
100/100 [==============================] - 23s 230ms/step - loss: 106.4602
Epoch 8/100
100/100 [==============================] - 24s 239ms/step - loss: 172.0166
Epoch 9/100
100/100 [==============================] - 23s 233ms/step - loss: 276.3909
Epoch 10/100
100/100 [==============================] - 23s 233ms/step - loss: 149.7979
Epoch 11/100
100/100 [==============================] - 24s 243ms/step - loss: 342.2365
Epoch 12/100
100/100 [==============================] - 23s 233ms/step - loss: 103.6166
Epoch 13/100
100/100 [==============================] - 24s 239ms/step - loss: 103.6175
Epoch 14/100
100/100 [==============================] - 23s 232ms/step - loss: 103.7286
Epoch 15/100
100/100 [==============================] - 23s 232ms/step - loss: 336.2559
Epoch 16/100
100/100 [==============================] - 24s 239ms/step - loss: 103.6171
Epoch 17/100
100/100 [==============================] - 24s 235ms/step - loss: 103.6174
Epoch 18/100
100/100 [==============================] - 24s 239ms/step - loss: 103.6185
Epoch 19/100
100/100 [==============================] - 23s 230ms/step - loss: 103.6302
Epoch 20/100
100/100 [==============================] - 23s 231ms/step - loss: 131.1390
Epoch 21/100
100/100 [==============================] - 24s 240ms/step - loss: 103.6186

Please help me with this.

Thanks

**Comment by matpalm** [comment]
we use cosine distance when we're using the softmax formulation but for vanilla triplet loss things will be a _lot_ more stable if you use euclidean distance instead.

e.g. https://gist.github.com/matpalm/753b08c19f37d9c4234cd70b88e3c8bb

**Comment by matpalm** [comment]
as this isn't an issue against this repo i'm expecting someone will rightly close it. feel free to ask more on stack overflow.

**Comment by msafi04** [comment]
Thank you
